### PR TITLE
expression: make `builtinCastIntAsInt` support vectorized evaluation

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
Implement vectorized evaluation for `builtinCastIntAsInt`.

### What is changed and how it works?
1. Implement vectorized evaluation.
2. Add some tests.

Here is the benchmark in my MAC, almost 95x faster than before:
```
BenchmarkCastIntAsCastRow-12         200000              8542 ns/op               0 B/op          0 allocs/op
BenchmarkCastIntAsCastVec-12       20000000                90.9 ns/op             0 B/op          0 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_vectorized.go [hunk 1]
```diff
@@ -115,18 +115,24 @@ func (c *vecRowConverter) evalInt(row chunk.Row) (val int64, isNull bool, err er
...
 		return
 	}
 
 	input := row.Chunk()
-	oriSel := input.Sel()
-	input.SetSel(sel[:])
-	err = c.builtinFunc.vecEvalInt(input, buf)
-	input.SetSel(oriSel)
+	if input == nil {
+		// it happens when the user ensure the leaf node is a Constant and then put an empty Row:
```
Comment (qw4990) [inline expression/builtin_vectorized.go]: The way to handle this case now works but not graceful.
Would you have some other ideas to solve this problem?
@zz-jason @XuHuaiyu

File: expression/builtin_vectorized.go [hunk 2]
```diff
@@ -115,18 +115,24 @@ func (c *vecRowConverter) evalInt(row chunk.Row) (val int64, isNull bool, err er
...
+		//   ConstantExpr.EvalInt(chunk.Row{})
+		err = c.builtinFunc.vecEvalInt(chunk.VirChunk, buf)
```
Comment (XuHuaiyu) [inline expression/builtin_vectorized.go]: We may call `err = c.builtinFunc.vecEvalInt(nil, buf)` here.
And check whether input == nil in Constant.VecEvalXXX

Comment (qw4990) [reply expression/builtin_vectorized.go]: Good suggestion.
It has been updated.

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -435,15 +435,23 @@ func (b *builtinCastIntAsIntSig) Clone() builtinFunc {
...
+
+func (b *builtinCastIntAsIntSig) vecEvalInt(input *chunk.Chunk, result *chunk.Column) error {
+	if err := b.args[0].VecEvalInt(b.ctx, input, result); err != nil {
+		return err
 	}
-	if b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag) && res < 0 {
-		res = 0
+	if b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag) {
+		i64s := result.Int64s()
+		for i := range i64s {
```
Comment (zz-jason) [inline expression/builtin_cast.go]: it's better to add a comment about why we don't set null bitmap for `result`.

File: expression/builtin_vectorized.go [hunk 4]
```diff
@@ -115,18 +115,27 @@ func (c *vecRowConverter) evalInt(row chunk.Row) (val int64, isNull bool, err er
...
-	input.SetSel(sel[:])
-	err = c.builtinFunc.vecEvalInt(input, buf)
-	input.SetSel(oriSel)
+	if input == nil {
+		// it happens when the user ensure the leaf node is a Constant and then put an empty Row.
+		// for example, the expr is Plus(Constant1, Constant2), then expr.EvalInt(chunk.Row{}) is valid.
+		err = c.builtinFunc.vecEvalInt(nil, buf)
+	} else {
+		var sel [1]int
+		sel[0] = row.Idx()
```
Comment (zz-jason) [inline expression/builtin_vectorized.go]: how about:
```go
sel := []int{row.Idx()}
```

File: expression/builtin_vectorized.go [hunk 5]
```diff
@@ -115,18 +115,27 @@ func (c *vecRowConverter) evalInt(row chunk.Row) (val int64, isNull bool, err er
...
+		oriSel := input.Sel()
+		input.SetSel(sel[:])
+		err = c.builtinFunc.vecEvalInt(input, buf)
+		input.SetSel(oriSel)
+	}
+	if err != nil {
+		return 0, false, err
```
Comment (zz-jason) [inline expression/builtin_vectorized.go]:
```suggestion
		return 0, true, err
```

File: expression/builtin_cast_bench_test.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2017 PingCAP, Inc.
```
Comment (zz-jason) [inline expression/builtin_cast_bench_test.go]:
```suggestion
// Copyright 2019 PingCAP, Inc.
```

File: expression/builtin_cast_bench_test.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"testing"
+
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/mock"
+)
+
+func BenchmarkCastIntAsInt(b *testing.B) {
```
Comment (zz-jason) [inline expression/builtin_cast_bench_test.go]: could you post the benchmark code on the old cast implementation?

Comment (SunRunAway) [reply expression/builtin_cast_bench_test.go]: Try to use https://godoc.org/golang.org/x/perf/cmd/benchstat and post the benchmark result in the pull request description

File: expression/builtin_vectorized.go [hunk 8]
```diff
@@ -138,18 +147,25 @@ func (c *vecRowConverter) evalReal(row chunk.Row) (val float64, isNull bool, err
 	if !c.builtinFunc.vectorized() {
 		return c.builtinFunc.evalReal(row)
 	}
-	var sel [1]int
-	sel[0] = row.Idx()
 	var buf *chunk.Column
 	if buf, err = c.builtinFunc.get(types.ETInt, 1); err != nil {
```
Comment (SunRunAway) [inline expression/builtin_vectorized.go]: It's still`types.ETInt` here, is that correct?

File: expression/builtin_cast.go (hunk 3 above)
Comment (qw4990) [reply expression/builtin_cast.go]: Some comments are added.

File: expression/builtin_vectorized.go (hunk 8 above)
Comment (qw4990) [reply expression/builtin_vectorized.go]: Now the converter has been removed.

File: expression/builtin_cast_bench_test.go (hunk 7 above)
Comment (qw4990) [reply expression/builtin_cast_bench_test.go]: It is implemented here (https://github.com/pingcap/tidb/pull/11826/files#diff-2967d2754f43b7d6318020f0953a69f3R38)

**Pull request merged**